R/peaks-functions.R

```diff
@@ -125,3 +127,66 @@ NULL
     x <- x[x[, 2] > .qtl, 1]
     quantile(diff(x), 0.25) > k
 }
+
+#' @description Compare peaks between 2 spectra
+#'
+#' @description
+#'
+#' Compare peaks from one spectrum to peaks from another spectrum with the
+#' provided function (`FUN`). `.peaks_compare_intensities` ensures that peaks
+#' are first matched between the two spectra with the [closest()] function.
+#'
+#' @param x peaks `matrix` with columns `"mz"` and `"intensity"`.
+#'
+#' @param y peaks `matrix` with columns `"mz"` and `"intensity"`.
+#'
+#' @param FUN `function` to be applied to the intensity values of matched
+#'     peaks between peak lists `x` and `y`. The function should take two
+#'     `numeric` vectors as their first argument and should return a single
+#'     `numeric`. Parameter `...` will be passed to the function.
+#'
+#' @param tolerance `numeric(1)` with the acceptable (constant) difference in
+#'     m/z values for matching peaks.
+#'
+#' @param ppm `numeric(1)` specifying a peak-specific, relative,
+#'     tolerance for the peak matching. Peaks between the two spectra with a
+#'     difference in their m/z smaller than *ppm* of the m/z of the peak (of
+#'     the second spectrum) are matched. See also [ppm()].
+#'
+#' @param ... additional parameters passed to the `FUN` function.
+#'
+#' @return The result from `FUN`, which should be a `numeric(1)`.
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+#'
+#' @importFrom stats cor
+#'
+#' @importFrom MsCoreUtils closest ppm
+#'
+#' @examples
+#'
+#' x <- cbind(c(31.34, 34.43, 54.65, 300.12, 514.5, 856.12),
+#'      c(13, 15, 56, 7, 45, 321))
+#'
+#' y <- cbind(c(34.431, 35.34, 50.24, 300.121, 514.51, 856.122),
+#'      c(123, 34, 323, 432, 12, 433))
+#'
+#' Spectra:::.peaks_compare_intensities(x, y)
+#' Spectra:::.peaks_compare_intensities(x, y, ppm = 40)
+#' Spectra:::.peaks_compare_intensities(x, y, ppm = 0)
+#'
+#' ## To get the number of common peaks
+#' Spectra:::.peaks_compare_intensities(x, y, FUN = function(x, y) length(x))
+#' Spectra:::.peaks_compare_intensities(x, y, FUN = function(x, y) length(x), ppm = 40)
+#' ## NA if there are none common
+#' Spectra:::.peaks_compare_intensities(x, y, FUN = function(x, y) length(x), ppm = 0)
+.peaks_compare_intensities <- function(x, y, FUN = cor, tolerance = 0,
+                                       ppm = 20, ...) {
+    matches <- closest(x[, 1], y[, 1], tolerance = tolerance + ppm(y[, 1], ppm))
+    not_na <- !is.na(matches)
+    if (any(not_na)) {
+        FUN(x[not_na, 2], y[matches[not_na], 2], ...)
```

Comment:
I am not sure that this function really does what it is intended to do.
See this toy example with just 2 common peaks out of 20 gets a perfect `cor` of `1` with binning it looks completely different:
```r
x <- cbind(1:10, 1:10)
y <- cbind(9:20, 9:20)

.peaks_compare_intensities(x, y)
# [1] 1

xb <- bin(x[, 2], x[, 1], size=1, breaks=1:20)
yb <- bin(y[, 2], y[, 1], size=1, breaks=1:20)
xb <- cbind(xb$mids, xb$x)
yb <- cbind(yb$mids, yb$x)

.peaks_compare_intensities(xb, yb)
# [1] -0.552357
```

Reply:
That's a good point. We should discuss this also in issue #46 - and eventually involve people that do actually MS2 spectra comparisons - I've never done that so far...

Reply:
Binning assumes that the peaks that are not there have an intensity of `0` - my gut feeling tells me that is also not OK.

Reply:
Code point to ask people that are more familiar with this specific topic. Let's wait what kind of input we get.